storage_controller/src/service.rs

```diff
@@ -4131,17 +4131,38 @@ impl Service {
         .ok_or_else(|| ApiError::NotFound(anyhow::anyhow!("Tenant {tenant_id} not found").into()))
     }
 
-    pub(crate) fn tenant_list(&self) -> Vec<TenantDescribeResponse> {
+    /// limit & offset are pagination parameters. Since we are walking an in-memory HashMap, `offset` does not
+    /// avoid traversing data, it just avoid returning it. This is suitable for our purposes, since our in memory
+    /// maps are small enough to traverse fast, our pagination is just to avoid serializing huge JSON responses
+    /// in our external API.
+    pub(crate) fn tenant_list(
+        &self,
+        limit: Option<usize>,
+        mut offset: usize,
```

Comment:
I think it would be better to use a tenant ID for the offset. the tenants are stored in a btree map, so it should be cheap to seek to the specified tenant.

in the current implementation, listing all tenants is an `O(n^2)` operation (assuming constant limit). furthermore, any additional tenants that got added before the offset lead to some tenants being listed multiple times (in addition to the added ones not being listed, but i think that's okay).

Reply:
I'm not too concerned about the runtime (serialization is expensive, walking the list is very cheap in comparison), however when I think about the edge case of handling responses when items are added/inserted, using offsets risks omitting other tenants when something is removed, whereas using tenant IDs guarantees that we won't do that, so it's preferable.

I've updated the PR to use a `start_after` parameter that is a tenant ID.